### PR TITLE
executor: improve wide table insert & update performance

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

from one of our customers, wide table's INSERT in 2.1 is slower than 2.0

the question is `evalRow` call `chunk.MutRowFromDatums` values-len x column-num times, and `MutRowFromDatums` will make heavy alloc operation

this is cpu pprof result in 58 column insert under dead loop

x

### What is changed and how it works?

eval buffer can be reuse and modify partly(if `x = 3, y = x + 1` situation)

### Check List 

Tests 

 - Unit test
 - Manual test compare to old code

single thread dead loop to execute a sql `insert into 58 columns table` and monitor `insertValues#insertRows` time.

before this PR:

```
====> 13.343243ms
====> 13.249525ms
====> 12.252023ms
====> 21.257291ms
====> 19.790786ms
====> 12.445035ms
====> 23.729567ms
====> 11.301022ms
====> 12.192163ms
====> 12.253682ms
====> 11.215405ms
====> 11.362502ms
```

after this PR:

```
----> 579.105µs
----> 589.978µs
----> 647.785µs
----> 1.184076ms
----> 678.121µs
----> 1.005605ms
----> 565.764µs
----> 617.942µs
----> 628.033µs
----> 569.044µs
----> 575.293µs
----> 605.097µs
----> 569.361µs
----> 935.86µs
```

Code changes

 - impl change

Side effects

 - no

Related changes

 - Need to cherry-pick to 2.1
 - Need to be included in the release note

Other Question

 `MutRowFromDatums` is devil, we should avoid to use it  and delete other usage too

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/7935)

### CODE REVIEW

**Comment by ngaut** [comment]
Please add a performance test to integrate tests.

**Comment by lysu** [comment]
There are some bugs in first version modification.

it works well if `expr` is `const` or some expression that doesn't do unsafe operation on chunk.

but when `expr` is sth like `column` will meet data corruption, `eval` will unsafe reuse `evalBuffer` memory and update wrong data.

so Now, we add `Flag()` method to `expression`(use flag because it maybe can provide more information like `strickness` in furtue) and `builtinFunc`....the prefect way is modify all builtinFunc impl flag, but it's too much work to do, so now just modify `baseBuilitnFuc`.

```
a expression is not reuse chunk memory if all it's children doesn't reuse chunk memory
```

so, after this PR, sql like `insert t values ('1', 2, 2, now())` or `update t set a = 1, b =2 where id =1` will be better, but sql like `update t1, t2 set t1.v = t2.v where t1.id = t2.id` will do like old code.

@zz-jason  @tiancaiamao  PTAL

File: executor/insert_common.go [hunk 1]
```diff
@@ -216,6 +240,13 @@ func (e *InsertValues) evalRow(cols []*table.Column, list []expression.Expressio
 
 		offset := cols[i].Offset
 		row[offset], hasValue[offset] = val1, true
+		if expr.Flag()&expression.FlagChunkReused > 0 {
```
Comment (tiancaiamao) [inline executor/insert_common.go]: The `FlagChunkReused` makes me confused, if the expression have a reuse flag, than we create a new chunk ?

Comment (lysu) [reply executor/insert_common.go]: yes, expression reuse chunk, so we can not reuse old chunk for following eval~

Comment (tiancaiamao) [reply executor/insert_common.go]: How about `expression.HoldChunkMemory`, it the returned result of an expression need to hold the memory of the chunk, we need to allocate a new chunk.

File: executor/insert_common.go [hunk 2]
```diff
@@ -117,9 +119,30 @@ func (e *InsertValues) getColumns(tableCols []*table.Column) ([]*table.Column, e
...
+	e.initEvalBuffer()
+
 	return cols, nil
 }
 
+func (e *InsertValues) initEvalBuffer() {
+	if e.evalBuffer != nil {
+		return
+	}
+	tpsLen := len(e.Table.Cols())
```
Comment (zz-jason) [inline executor/insert_common.go]: how about s/tpsLen/numCols/

File: executor/insert_common.go [hunk 3]
```diff
@@ -117,9 +119,30 @@ func (e *InsertValues) getColumns(tableCols []*table.Column) ([]*table.Column, e
...
+		tpsLen++
+	}
+	e.evalBufferTypes = make([]*types.FieldType, tpsLen)
+	for i, col := range e.Table.Cols() {
+		e.evalBufferTypes[i] = &col.FieldType
+	}
+	if e.hasExtraHandle {
+		e.evalBufferTypes[len(e.evalBufferTypes)-1] = types.NewFieldType(mysql.TypeLonglong)
+	}
+	mutChunk := chunk.MutRowFromTypes(e.evalBufferTypes)
```
Comment (zz-jason) [inline executor/insert_common.go]: how about s/mutChunk/mutRow/

File: expression/expression.go [hunk 4]
```diff
@@ -36,6 +36,14 @@ const (
 	scalarFunctionFlag byte = 3
 )
 
+// Flag represents expression's property.
+type Flag byte
+
+const (
+	// FlagHoldChunkMemory indicates expression return result maybe returned by unsafe operation over chunk.
+	FlagHoldChunkMemory Flag = 1 << iota
```
Comment (zz-jason) [inline expression/expression.go]: By default, this is true. I think we can remove this flag and the `Flag()` API from the `expression` interface.

File: executor/insert_common.go [hunk 5]
```diff
@@ -117,9 +119,30 @@ func (e *InsertValues) getColumns(tableCols []*table.Column) ([]*table.Column, e
 		return nil, errors.Trace(err)
 	}
 
+	e.initEvalBuffer()
```
Comment (zz-jason) [inline executor/insert_common.go]: This function call can be put to `InsertExec.Open()`?

File: expression/expression.go [hunk 6]
```diff
@@ -36,6 +36,14 @@ const (
 	scalarFunctionFlag byte = 3
 )
 
+// Flag represents expression's property.
+type Flag byte
```
[the remark this answers is not in the document]
Comment (zhexuany) [reply expression/expression.go]: It seems to be used in next few lines. ``` FlagHoldChunkMemory Flag = 1 << iota ```

[the remark this answers is not in the document]
Comment (zz-jason) [reply expression/expression.go]: Yeah, I mean this line and the flowing few lines can all be removed.

File: executor/insert_common.go (hunk 5 above)
Comment (zz-jason) [reply executor/insert_common.go]: `getColumns()` is called in the `Next()` function, Maybe we can move it to the `Open()` function:
```go
InsertExec.Open:
	getColumns()
	initEvalBuffer()

InsertExec.Next:
	e.insertRowsFromSelect(), or:
	e.insertRows()
```

Comment (zz-jason) [reply executor/insert_common.go]: another thing is, according to the callers of the function `getColumns()`, maybe we can also improve the performance of the `replace` statement:
```
  3 executor/insert.go|132 col 17| cols, err := e.getColumns(e.Table.Cols())
  4 executor/replace.go|181 col 17| cols, err := e.getColumns(e.Table.Cols())
  5 executor/builder.go|576 col 28| columns, err := insertVal.getColumns(tableCols)
```

Comment (lysu) [reply executor/insert_common.go]: finally, I move getColumns to buildExecutor just like  loadData does...and move initEvalBuffer to Open, and replace pointer to direct ref.

PTAL @zz-jason , although CI doesn't work

**Comment by tiancaiamao** [comment]
LGTM @zz-jason

File: executor/insert_common.go [hunk 7]
```diff
@@ -50,9 +50,13 @@ type InsertValues struct {
 	GenColumns []*ast.ColumnName
 	GenExprs   []expression.Expression
 
+	InsertColumns []*table.Column
```
Comment (zz-jason) [inline executor/insert_common.go]: no need to export it? s/InsertColumns/insertColumns/

File: executor/insert_common.go [hunk 8]
```diff
@@ -114,10 +120,25 @@ func (e *InsertValues) getColumns(tableCols []*table.Column) ([]*table.Column, e
 	// Check column whether is specified only once.
 	err = table.CheckOnce(cols)
 	if err != nil {
-		return nil, errors.Trace(err)
+		return errors.Trace(err)
```
Comment (zz-jason) [inline executor/insert_common.go]: no need to `Trace` any more?

**Pull request merged**